modules/widgets/class-jetpack-instagram-widget.php

```diff
@@ -0,0 +1,568 @@
+			return array( 'valid' => false );
+		}
+		if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+			$token = Keyring::init()->get_token_store()->get_token(
+				array(
+					'type' => 'access',
+					'id'   => $token_id,
+				)
+			);
+
+			return array(
+				'valid'  => ! empty( $token ),
+				'legacy' => $token && 'instagram' === $token->name,
+			);
+		}
+
+		$site          = Jetpack_Options::get_option( 'id' );
+		$path          = sprintf( '/sites/%s/instagram/%s/check-token', $site, $token_id );
+		$result        = Client::wpcom_json_api_request_as_blog( $path, 2, array( 'headers' => array( 'content-type' => 'application/json' ) ), null, 'wpcom' );
+		$response_code = wp_remote_retrieve_response_code( $result );
+		if ( 200 !== $response_code ) {
+			return array(
+				// We assume the token is valid if the response_code is anything but the invalid
+				// token codes we send back. This is to make sure it's not reset, if the API is down
+				// or something.
+				'valid'  => ! ( 403 === $response_code || 401 === $response_code ),
+				'legacy' => 'ERROR',
+			);
+		}
+		$status = json_decode( $result['body'], true );
+		return $status;
+	}
+
+	/**
+	 * Validates the widget instance's token ID and then uses it to fetch images from Instagram.
+	 * It then caches the result which it will use on subsequent pageviews.
+	 * Keyring is not loaded nor is a remote request is not made in the event of a cache hit.
+	 *
+	 * @param array $instance A widget $instance, as passed to a widget's widget() method.
+	 * @return WP_Error|array A WP_Error on error, an array of images on success.
+	 */
+	public function get_data( $instance ) {
+		if ( empty( $instance['token_id'] ) ) {
+			return new WP_Error( 'empty_token', 'The token id was empty', 403 );
+		}
+
+		$cache_time    = MINUTE_IN_SECONDS;
+		$transient_key = implode( '|', array( 'instagram-widget', $instance['token_id'], $instance['count'] ) );
+		$cached_images = get_transient( $transient_key );
+		if ( $cached_images ) {
+			return $cached_images;
+		}
+
+		$site   = Jetpack_Options::get_option( 'id' );
+		$path   = sprintf( '/sites/%s/instagram/%s?count=%s', $site, $instance['token_id'], $instance['count'] );
+		$result = Client::wpcom_json_api_request_as_blog( $path, 2, array( 'headers' => array( 'content-type' => 'application/json' ) ), null, 'wpcom' );
+
+		$response_code = wp_remote_retrieve_response_code( $result );
+		if ( 200 !== $response_code ) {
+			set_transient( $transient_key, 'ERROR', $cache_time );
+			return new WP_Error( 'invalid_response', 'The response was invalid', $response_code );
+		}
+
+		$data = json_decode( wp_remote_retrieve_body( $result ), true );
+		if ( ! isset( $data['images'] ) || ! is_array( $data['images'] ) ) {
+			set_transient( $transient_key, 'ERROR', $cache_time );
+			return new WP_Error( 'missing_images', 'The images were missing', $response_code );
+		}
+
+		$cache_time = 20 * MINUTE_IN_SECONDS;
+		set_transient( $transient_key, $data, $cache_time );
```

Comment:
For the block, we cache things for an hour, and not at all when we have errors. should we do the same here? 
https://github.com/Automattic/jetpack/blob/12e2c624ebbc58cd03b8b1ea3ccbcc29a4e2d3b8/_inc/lib/class-jetpack-instagram-gallery-helper.php#L98

We can also make the error messages translatable.

I think we could also prefix the transient name with `jetpack_`. It would make it easier for folks and for us when we want to delete all traces of Jetpack from a site.

```suggestion
		$transient_key = implode( '|', array( 'jetpack_instagram_widget', $instance['token_id'], $instance['count'] ) );
		$cached_images = get_transient( $transient_key );
		if ( $cached_images ) {
			return $cached_images;
		}

		$site   = Jetpack_Options::get_option( 'id' );
		$path   = sprintf( '/sites/%s/instagram/%s?count=%s', $site, $instance['token_id'], $instance['count'] );
		$result = Client::wpcom_json_api_request_as_blog( $path, 2, array( 'headers' => array( 'content-type' => 'application/json' ) ), null, 'wpcom' );

		$response_code = wp_remote_retrieve_response_code( $result );
		if ( 200 !== $response_code ) {
			return new WP_Error( 'invalid_response', esc_html__( 'The response was invalid', 'jetpack' ), $response_code );
		}

		$data = json_decode( wp_remote_retrieve_body( $result ), true );
		if ( ! isset( $data['images'] ) || ! is_array( $data['images'] ) ) {
			return new WP_Error( 'missing_images', esc_html__( 'The images were missing', 'jetpack' ), $response_code );
		}

		set_transient( $transient_key, $data, HOUR_IN_SECONDS );
```